src/MultiFormats/MultiAddress.cs

```diff
@@ -0,0 +1,565 @@
+            : this()
+        {
+            var type = ip.AddressFamily == AddressFamily.InterNetwork
+                ? "ip4"
+                : "ip6";
+            Read(new StringReader($"/{type}/{ip}"));
+        }
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="MultiAddress"/> class from the
+        ///   specified <see cref="IPEndPoint"/>.
+        /// </summary>
+        public MultiAddress(IPEndPoint endpoint)
+            : this()
+        {
+            var type = endpoint.AddressFamily == AddressFamily.InterNetwork
+                ? "ip4"
+                : "ip6";
+            Read(new StringReader($"/{type}/{endpoint.Address}/tcp/{endpoint.Port}"));
+        }
+
+        /// <summary>
+        ///   Creates a deep copy of the multi address.
+        /// </summary>
+        /// <returns>
+        ///   A new deep copy.
+        /// </returns>
+        public MultiAddress Clone() { return new MultiAddress(ToString()); }
+
+        /// <summary>
+        ///   Creates a new instance of the <see cref="MultiAddress"/> class from the
+        ///   specified byte array.
+        /// </summary>
+        /// <param name="buffer">(
+        ///   A byte array containing the binary representation of a
+        ///   <b>MultiAddress</b>.
+        /// </param>
+        /// <remarks>
+        ///   Reads the binary representation of <see cref="MultiAddress"/> from the <paramref name="buffer"/>.
+        ///   <para>
+        ///   The binary representation is a sequence of <see cref="NetworkProtocol">network protocols</see>.
+        ///   </para>
+        /// </remarks>
+        public MultiAddress(byte[] buffer)
+            : this()
+        {
+            if (buffer == null || buffer.Length == 0)
+            {
+                return;
+            }
+
+            Read(new MemoryStream(buffer, false));
+        }
+
+        /// <summary>
+        ///   Gets the peer ID of the multiaddress.
+        /// </summary>
+        /// <value>
+        ///   The <see>
+        ///       <cref>Lib.P2P.Peer.Id</cref>
+        ///   </see>
+        ///   as a <see cref="MultiHash"/>.
+        /// </value>
+        /// <exception cref="Exception">
+        ///   When the last <see cref="Protocols">protocol</see>
+        ///   is not "ipfs" nor "p2p".
+        /// </exception>
+        /// <remarks>
+        ///   The peer ID is contained in the last protocol that
+        ///   is "ipfs" or "p2p".  For example, <c>/ip4/10.1.10.10/tcp/29087/ipfs/QmVcSqVEsvm5RR9mBLjwpb2XjFVn5bPdPL69mL8PH45pPC</c>.
+        /// </remarks>
+        public MultiHash PeerId
+        {
+            get
+            {
+                var protocol = Protocols
+                   .LastOrDefault(p => p.Name == "ipfs" || p.Name == "p2p");
+                if (protocol == null)
+                {
+                    throw new Exception($"'{this}' is missing the peer ID. Add the 'ipfs' or 'p2p' protocol.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4979568)